bridge/ffi/src/c_api.rs

```diff
@@ -232,54 +360,86 @@ pub unsafe extern "C" fn graph_get_connections_without_keys(
+			Ok(connections) => {
+				let connections_len = connections.len();
+				let connections_ptr = ManuallyDrop::new(connections).as_mut_ptr();
+				let graph_connections =
+					GraphConnections { connections: connections_ptr, connections_len };
+				DsnpGraphFFIResult::new(graph_connections)
+			},
+			Err(error) =>
+				DsnpGraphFFIResult::new_mut_error(DsnpGraphErrorHandle::from_error(error)),
+		}
 	});
-	result.unwrap_or(GraphConnections { connections: std::ptr::null_mut(), connections_len: 0 })
+	result.unwrap_or_else(|error| {
+		DsnpGraphFFIResult::new_mut_error(DsnpGraphErrorHandle::from_error(
+			DsnpGraphError::Unknown(anyhow::anyhow!(
+				"Failed get one sided private friendship connections from graph: {:?}",
+				error
+			)),
+		))
+	})
 }
 
 // Get a list published and imported public keys associated with a user
 #[no_mangle]
 pub unsafe extern "C" fn graph_get_public_keys(
 	graph_state: *mut GraphState,
 	user_id: *const DsnpUserId,
-) -> DsnpPublicKeys {
+) -> DsnpGraphFFIResult<DsnpPublicKeys, DsnpGraphErrorHandle> {
 	let result = panic::catch_unwind(|| {
 		if graph_state.is_null() {
-			return DsnpPublicKeys { keys: std::ptr::null_mut(), keys_len: 0 }
+			return DsnpGraphFFIResult::new_mut_error(DsnpGraphErrorHandle::from_error(
+				DsnpGraphError::FFIError("Graph state is null".to_string()),
+			))
 		}
 		let graph_state = &mut *graph_state;
 		let user_id = &*user_id;
-		let keys = graph_state.get_public_keys(user_id).unwrap_or_default();
-		let ffi_keys = dsnp_public_keys_to_ffi(keys);
-		let keys_len = ffi_keys.len();
-		let keys_ptr = ManuallyDrop::new(ffi_keys).as_mut_ptr();
-		DsnpPublicKeys { keys: keys_ptr, keys_len }
+
+		match graph_state.get_public_keys(user_id) {
+			Ok(keys) => {
+				let ffi_keys = dsnp_public_keys_to_ffi(keys);
+				let keys_len = ffi_keys.len();
+				let keys_ptr = ManuallyDrop::new(ffi_keys).as_mut_ptr();
+				let public_keys = DsnpPublicKeys { keys: keys_ptr, keys_len };
+				DsnpGraphFFIResult::new(public_keys)
+			},
+			Err(error) =>
+				DsnpGraphFFIResult::new_mut_error(DsnpGraphErrorHandle::from_error(error)),
+		}
 	});
-	result.unwrap_or(DsnpPublicKeys { keys: std::ptr::null_mut(), keys_len: 0 })
+	result.unwrap_or_else(|error| {
+		DsnpGraphFFIResult::new_mut_error(DsnpGraphErrorHandle::from_error(
+			DsnpGraphError::Unknown(anyhow::anyhow!(
+				"Failed get public keys from graph: {:?}",
+				error
+			)),
+		))
+	})
 }
 
-// free graph state
+// Free graph state
 #[no_mangle]
 pub unsafe extern "C" fn free_graph_state(graph_state: *mut GraphState) {
 	let result = panic::catch_unwind(|| {
 		if graph_state.is_null() {
 			return
 		}
 		let mut graph_states = GRAPH_STATES.lock().unwrap();
-		let index = graph_states.iter().position(|x| x as *const _ == graph_state).unwrap();
-		graph_states.remove(index);
+		let index =
+			graph_states.iter().position(|x| x.as_ref() as *const _ == graph_state).unwrap();
+		let _ = ManuallyDrop::new(graph_states.remove(index));
```

Comment:
Yes it should, but i wanted it to be more predictable and wanting to drop in the code exclusively then depending on vec  to do it for us, thought this is cleaner and easy to understand given this is where we want to be very sure given its the state.

do you prefer otherwise?

Reply:
I'm just worried about double-free than anything. I'm more leaning towards depending on Rust for managing the memory as much as possible and if there are cases that we must handle it manually then it's fine

Reply:
Yeah no worries just pushed an update, letting rust do its thingy